Muzero

2xic opened the pull request:
The code is not done yet. Some of the big things missing are

- ~~the full loss is not calculated (missing policy and value function)~~ (done)
- ~~no replay buffer -> only last game state is used currently used for testing.  (replay buffer should not be necessary for the basic environment)~~
- ~~no self play yet (should not be necerrary for the basic environment)~~
- not using residual blocks and convolution (basic change, but not needed for the basic environment)

I added a basic environment to make testing easy (see `basic_env.py` in utils). When the model is able to "win" in that environment then this PR will be ready to be merged.  

I will also update this commit with a pep8 code style checker. Seems like pylint does not report about invalid code styles.

2xic committed:
basic tree setup

2xic committed:
scaffolding for muzero

Zikoat committed:
WIP: Created Guided minesweeper environment

Zikoat committed:
Add MrGris code to PyCharm

2xic committed:
full loss implemened

2xic committed:
removed files

2xic committed:
basic replay buffer

2xic committed:
run on gpu if cuda is available

2xic committed:
fixed bug

2xic committed:
added ucb

2xic committed:
nosie should be added when exploring

2xic committed:
trying to debug why training takes so long

2xic commented on musweeper/musweeper/muzero/tree/monte_carlo.py:
```diff
@@ -0,0 +1,201 @@
+			if (should_pick_random_node or current_node_has_no_children) and len(current_node.children) != self.children_count:
+				# TODO : add something to invalidate invalid states to make search easier when bootstraping
+				new_node = current_node.get_a_children_node(self.children_count)
+				# TODO : you should actually backpropgate as soon as a leaf is found
```
remove comment. this is now one

2xic commented on musweeper/musweeper/muzero/tree/monte_carlo.py:
```diff
@@ -0,0 +1,201 @@
+		current_node = self.root if current_node is None else current_node
+		found_leaf = False
+		current_node.add_exploration_noise()
+		# TODO : add "or" for if current node is a terminated state 
```
remove

2xic commented on musweeper/musweeper/muzero/tree/node.py:
```diff
@@ -0,0 +1,130 @@
+		self.min = 0
+
+#	@staticmethod
+#	def calulate_q(self):
```
remove, also the min - max node should be called when a node is created.

2xic commented on musweeper/musweeper/muzero/tree/node.py:
```diff
@@ -0,0 +1,130 @@
+
+	def search_value_exploration_exploration(self):
+		# https://en.wikipedia.org/wiki/Monte_Carlo_tree_search
+		# TODO : Muzero uses a more complex version, change this to the one from the paper
```
remove- this is now implemented.

UCB should be better than this function. So make UCB standard.

2xic commented on musweeper/musweeper/muzero/tree/node.py:
```diff
@@ -0,0 +1,130 @@
+		return child_explored + c * np.sqrt(parrent_explored)
+
+	def upper_confidente_boundary(self):
+		self.c1 = 1.25
```
add docstrings

2xic commented on musweeper/musweeper/muzero/utils/game_history.py:
```diff
@@ -0,0 +1,32 @@
+	def length(self):
+		return len(self.state)
+
+class reply_buffer:
```
replay

2xic commented on musweeper/musweeper/muzero/utils/game_history.py:
```diff
@@ -0,0 +1,32 @@
+import random
+
+class game_history:
+	def __init__(self):
```
one of the variables are used for rollout and one for history. So you should rewrite this.

2xic reviewed (commented):
fix this tomorrow

- test the model in some other gym enviorment (cartpole or something)

- ~~since there is no "player" in the minesweeper environment. Only single player. The self play mechanism has to be based on optimal rewards in the game. Play the game x times, sort game based on total score, the lowest score is loosing players and highest score is winning player~~.  

- also the model spends a lot of time computing when retraining. We are recomputing the history when doing the rollout. I think we have to do it this way, but reread the paper and see if you missed some trick. Also, why does it take so long time when we are using cuda ?

2xic committed:
min-max value function calculation

2xic committed:
correctly implement backpropagate of monte carlo search tree

2xic committed:
fixed bug

2xic committed:
rewrote game history

2xic committed:
pseudo code for self-play

2xic committed:
finished selfplay with tests and added l2 regularization to optimizer

Zikoat committed:
Implement calling python2.7 code from python3

Zikoat committed:
Correctly register guided environment

Zikoat committed:
Decode and evaluate api_solve output, fix appending matrix to obs space

Zikoat committed:
Merge branch 'main' into feature/guided-env

Zikoat committed:
Added test solving an ascii board

Zikoat committed:
Create probabliity matrix out of solution

Zikoat committed:
WIP: Move probability matrix to guided env

2xic committed:
more tests, fixed ucb calculation bug and basic debugger. Still not learning optimally

2xic committed:
smaller model, longer search time and longer training time. Will check new results tomorrow

Zikoat committed:
Fix error when creating boards of unequal width and height

Zikoat committed:
Fix rest of GuidedEnv tests, better logging on python2.7 code

Zikoat committed:
Standardized env metadata

Zikoat committed:
Fixed array size and added docs to environment.

2xic committed:
fixed ucb nan problem, more tests and asserts

2xic committed:
prior was not correctly set. still not fully fixed.

2xic committed:
prior is now set when node is expanded

2xic committed:
fixed 'sticky' actions

2xic committed:
some bug fixes, code for testing with minesweeper with reward hacking

2xic committed:
removed duplicate code and added node pruning

Zikoat commented on musweeper/musweeper/muzero/tree/monte_carlo.py:
```diff
@@ -0,0 +1,259 @@
+            delta_depth = (output_node.depth - node.depth)
+            self.backpropgate(output_node, delta_depth)
+            visited_nodes += 1
+        return output_node
```
The code in model.plan_action, tree.expand_node and node.expand doesnt really follow the pseudocode. 

When we run a simulation through the MCTS, we should start once at the root, then follow the path down the tree with the node that has the largest UCB and when we are at the bottom, we expand the node and backpropagate new values. We then restart the search from the top. 

It seems like what you are doing here is expanding every child of every node max_search_depth deep, but in the correct algorithm we do not care how deep down in the tree we are, we only care about how many leaf nodes we have visited.

2xic committed:
fixed search tree bug

Zikoat committed:
Remove unused code

Zikoat committed:
Fix probability calculation when _other is used

and other small fixes

Zikoat committed:
Add seeding and other fixes

Zikoat committed:
Rename _ to info

Zikoat committed:
Add assisted environment

Zikoat committed:
Add description to guided env, and fix nondeterministic setting

Zikoat committed:
Moved invariants to own function

Zikoat committed:
Add legal_action

Zikoat committed:
Add todos

Zikoat committed:
WIP: start work on mrgris agent

Zikoat committed:
Add legal_actions for guided and assisted environment

Zikoat committed:
Massive cleanup dump 1

Zikoat committed:
Massive Refactor

Using pycharm "refarmat code" and going through warnings. Not all the warnings are gone, but the most severe are removed. 

I just wanna welcome all bisect users here.

2xic committed:
added temperature and fixed search tree

Zikoat committed:
Refactored _get_info

Zikoat committed:
Fix guided env reset doesnt return probabilities

Zikoat committed:
WIP: Add mrgris agent

2xic committed:
added tensorboard logger and debugger for model

2xic committed:
fixed loss - should come from model not monte carlo tree

2xic committed:
more bug fixes for loss

2xic committed:
will naive search beat MCTS ?

2xic committed:
policy should reflect the MCTS search tree

2xic committed:
debug code

2xic committed:
Only do legal actions from root

2xic committed:
fixed legal action filter bug

2xic committed:
conv blocks for representation and make action one-hot vector

2xic committed:
agent evaluation code from master

2xic committed:
just some changes for training and eval

2xic committed:
merged in changes from feature/guided-env

2xic committed:
added mrgris python3 code

2xic committed:
fixed mrgris code and added mrgris probability_matrix to loss

2xic committed:
model evaluation files

2xic committed:
added missing tests

2xic committed:
fixed typo

2xic committed:
docstrings and remvoed unused code
